ui-v2/app/initializers/nspace.js

```diff
@@ -0,0 +1,61 @@
+import Route from '@ember/routing/route';
+import config from 'consul-ui/config/environment';
+
+import { routes } from 'consul-ui/router';
+import flat from 'flat';
+
+let initialize = function() {};
+Route.reopen(
+  ['modelFor', 'transitionTo', 'paramsFor'].reduce(function(prev, item) {
+    prev[item] = function(routeName, ...rest) {
+      const isNspaced = this.routeName.startsWith('nspace.');
+      if (routeName === 'nspace') {
+        if (isNspaced || this.routeName === 'nspace') {
+          return this._super(...[routeName, ...rest]);
+        } else {
+          return {
+            nspace: '~',
+          };
+        }
+      }
+      if (isNspaced && routeName.startsWith('dc')) {
+        routeName = `nspace.${routeName}`;
+      }
+      return this._super(...[routeName, ...rest]);
+    };
+    return prev;
+  }, {})
+);
+if (config.CONSUL_NSPACES_ENABLED) {
+  const dotRe = /\./g;
+  initialize = function(container) {
+    const all = Object.keys(flat(routes))
+      .filter(function(item) {
+        return item.startsWith('dc');
+      })
+      .map(function(item) {
+        return item.replace('._options.path', '').replace(dotRe, '/');
+      });
+    all.forEach(function(item) {
+      let route = container.resolveRegistration(`route:${item}`);
+      if (!route) {
+        item = `${item}/index`;
+        route = container.resolveRegistration(`route:${item}`);
+      }
+      route.reopen({
+        templateName: item
+          .replace('/root-create', '/create')
+          .replace('/create', '/edit')
+          .replace('/folder', '/index'),
+      });
```

Comment:
It would be great if we could read the `templateName` property on the Route before reopening it. Kind of like `const currentTemplate = Route.prototype.templateName`. If anyone knows how to do that I'd be very grateful. Otherwise I'm guessing once we move to Octane/JS classes we should be able to just inspect the prototype.

Reply:
I think you may be able to do `route.proto()` to get a reference to the prototype.